dpkg-tmplocal.8
.TH dpkg-tmplocal 8 "Release 1.0" "LLNL" "dpkg-tmplocal"
.SH NAME
dpkg-tmplocal \- run process with private copy of /usr/local
.SH SYNOPSIS
.B dpkg-tmplocal
.I "[-v|-p]"
.I "[-d directory]"
.I "cmd [arg] ..."
.SH DESCRIPTION
.B dpkg-tmplocal
is part of dpkg-scripts, a collection of utilities for building
/usr/local-based deb packages.
.LP
.B dpkg-tmplocal
creates a unique, empty temporary directory owned by the invoking user,
binds it over the top of /usr/local in a private namespace,
and runs the specified command in that environment.
With no options, the environment and the temporary directory are 
destroyed when the command terminates.
.LP
.B dpkg-tmplocal's
default behavior can be modified as follows:
.TP
\fI-v\fR, \fI--verbose\fR
Report each step as it is performed.
.TP
\fI-p\fR, \fI--preserve\fR
Do not remove the temporary directory after the command completes.
Use \fI--verbose\fR to see the name of the temporary directory as it is created.
.TP
\fI-d\fR, \fI--dir\fR directory
Bind a pre-existing directory on /usr/local.  Implies \fI--preserve\fR.
.TP
\fI-o\fR, \fI--slash-opt\fR
Bind on /opt instead of /usr/local.
.SH APPLICATIONS
.B dpkg-tmplocal
was originally designed as a helper for 
.B dpkg-mkdeb
which uses it during package creation when the software being
packaged must be installed directly into /usr/local instead of to an 
intermediate directory.  It also builds packages in a temporary 
local popluated only its PKG_BUILDREQUIRES packages to ensure they
are properly specified.
.LP
.B dpkg-tmplocal
is used by 
.B dpkg-testdeb
for running a package's unit tests in a temporary /usr/local containing
only its PKG_DEPENDS packages.
.LP
Finally,
.B dpkg-tmplocal
may be used to create a sandbox /usr/local for manual testing and development
of packages.  For example:
.IP
.nf
dpkg-tmplocal /bin/bash
$ dpkg-tmplocal -v /bin/bash
dpkg-tmplocal: binding /tmp/tmplocal-jPK0OB on /usr/local
dpkg-tmplocal: dropping root privileges
dpkg-tmplocal: execing /bin/bash
$ dpkg-initialize
$ dpkg -i --force-not-root pkg.deb
$ ...
$ exit
dpkg-tmplocal: removing /tmp/tmplocal-jPK0OB
.fi
.SH "SEE ALSO"
dpkg-mkdeb(8), dpkg-testdeb(8)


NEWS
========================================================================
Release Notes for dpkg-scripts-1.66                          17 Mar 2009
========================================================================

* Fix dpkg-tmplocal option processing to stop at the first non-option
  argument.

========================================================================
Release Notes for dpkg-scripts-1.65                          16 Mar 2009
========================================================================

* New dpkg-tmplocal options:
   -v (--verbose)   Report each step as it is performed.
   -p (--preserve)  Do not clean up temporary directory.
   -d (--dir)       Bind a pre-existing directory on /usr/local (implies -p)
   -o (--slash-opt) Bind on /opt instead of /usr/local.

* Avoid overriding PATH in dpkg-wrap [chaos bug 454]

========================================================================
Release Notes for dpkg-scripts-1.64                          22 Apr 2008
========================================================================

* Do not hide default package dotkits from "use -l" by default.

* Add 'dkhidedefault' package flag to hide default package dotkit.

* Add 'dkhidemain' package flag to hide main package dotkit.
  The 'dkhide' package flag hides both main and default packaged dotkits.

========================================================================
Release Notes for dpkg-scripts-1.63                          20 Apr 2008
========================================================================

* Fix bug in linkdata2 generation by dpkg-mkdeb where long symlink 
  targets were encoded into base64 with line breaks.

* Rewrite dpkg-verify in C for efficiency.  Work around linkdata2 files 
  damaged as described above.

* The dpkg-nonpackaged utility is replaced with dpkg-verify --unpackaged.

========================================================================
Release Notes for dpkg-scripts-1.62                          10 Apr 2008
========================================================================

* Added dpkg-mkdeb options: '-q basename', '-q basever', and -m metafile.
  Fixed a bug in dpkg-mkdeb -q default.

========================================================================
Release Notes for dpkg-scripts-1.61 (includes .59, .60)      09 Apr 2008
========================================================================

* Change '-q names' query to return only non-default packages names, 
  and add '-q default' package to return default package names.

* Add regression test.

========================================================================
Release Notes for dpkg-scripts-1.58                          20 Mar 2008
========================================================================

* Add 'verboselog' flag to allow logging of arguments passed to wrapped
  commands to be turned on for some packages.

========================================================================
Release Notes for dpkg-scripts-1.57                          28 Feb 2008
========================================================================

* Drop logging of arguments passed to wrapped commands.

========================================================================
Release Notes for dpkg-scripts-1.56                          21 Feb 2008
========================================================================

* Change hash-bang in all scripts to #!/bin/bash --posix to avoid
  high shell startup overhead due to LC BASH_ENV setting.
  Also make a couple of executable paths fully qualified in the 
  dpkg-wrap call chain to avoid gratuitous PATH walks.

* New dpkg-defaults script for listing package defaults.

========================================================================
Release Notes for dpkg-scripts-1.55                          15 Feb 2008
========================================================================

* Add dpkg-mkdeb -q names query to list all possible package names,
  needed on build farm.

* On failure, dpkg-mkdeb should not dump logfile to stderr if -v is 
  already sending it there.

* Do not build with 'notmproot' behavior by default.  Instead, populate
  a private /usr/local for PKG_BUILDREQUIRES, but install to a regular
  tmproot.  The -t option allows the system /usr/local to be used.

* Change the dpkg-docreport color scheme to white and gray.

========================================================================
Release Notes for dpkg-scripts-1.54                          12 Feb 2008
========================================================================

* Support for symlinking lcc plugins in default pacakges.

* New macros for use in package.conf: 
  $lccdir - location of lcc plugin directory
  $pkgname - package name (after subpackage processing)

* Ensure file permissions are such that deb can be installed by non-root
  user.

* Define $pkg_default (set to PKG_DEFAULT value) in package.macros
  metadata file.

* Handle some corner cases in PKG_DEFAULT parsing.

========================================================================
Release Notes for dpkg-scripts-1.53                          01 Feb 2008
========================================================================

* Add 'dkhide' flag to generate #a in dotkit, which hides it from
  the use -l listing.  Generate #a in default packages with or without
  'dkhide' flag.

* Add ${dotkitdir} macro which can be used instead of hard coding
  /usr/local/etc/dotkit in package.conf's.

* [dpkg-mkdeb] Make unknown entries in PKG_FLAGS fatal.
  
========================================================================
Release Notes for dpkg-scripts-1.52                          30 Jan 2008
========================================================================

* Add 'nocheckwrap' flag to disable checking for existence of wrapped 
  executables (e.g. to support wrapping of unpackaged files).


dpkg-tmplocal.c
/***************************************************************************
 Copyright (C) 2007 Lawrence Livermore National Security, LLC
 Produced at Lawrence Livermore National Laboratory.
 Written by Jim Garlick <garlick@llnl.gov>.
 UCRL-CODE-235516
 
 This file is part of dpkg-scripts, a set of utilities for managing 
 packages in /usr/local with dpkg.
 
 dpkg-scripts is free software; you can redistribute it and/or modify it 
 under the terms of the GNU General Public License as published by the Free
 Software Foundation; either version 2 of the License, or (at your option)
 any later version. 

 dpkg-scripts is distributed in the hope that it will be useful, but WITHOUT 
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for 
 more details.

 You should have received a copy of the GNU General Public License along
 with dpkg-scripts; if not, write to the Free Software Foundation, Inc.,
 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
***************************************************************************/

/* dpkg-tmplocal - bind a writeable tmpdir over /usr/local in private namespace 
 *
 * This file is installed setuid.  It does the following as root:
 * 1. creates tmp directory (chowned to the user)
 * 2. forks child process
 * 3. (child) unshares the namespace
 * 4. (child) binds tmp directory on /usr/local
 * The child process then loses its effective root id and runs the command as 
 * the real userid (which cannot be root).  Since the bind mount is 
 * performed in the private namespace of the child process, it is invisible 
 * to others and is automatically cleaned up when the child process exits.
 * The parent (still root) just waits for the child and removes the tmp
 * directory.
 */

#include <sys/types.h>
#include <sys/wait.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdlib.h>
#include <signal.h>
#include <sys/mount.h>
#include <sched.h>
#include <errno.h>
#include <libgen.h>
#include <string.h>
#include <grp.h>
#include <pwd.h>
#include <dirent.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <getopt.h>

static char 	*create_tmpdir(uid_t uid, gid_t gid);
static void	exec_cmd(uid_t uid, gid_t gid, int oopt,
			 int vopt, char *dir,char **argv);
static int	recursive_remove(char *name);

static char *prog;

#define OPTIONS "+od:vp"
static struct option longopts[] = {
    {"slash-opt",       no_argument,       0, 'o'},
    {"dir",             required_argument, 0, 'd'},
    {"verbose",         no_argument,       0, 'v'},
    {"preserve",        no_argument,       0, 'p'},
    {0, 0, 0, 0},
};

static void
usage(void)
{
	fprintf(stderr, "Usage: %s [-ovp] [-d dir] cmd [args...]\n", prog);
	exit(1);
}


int
main(int argc, char **argv)
{
	pid_t pid, p;
	int exit_code = 0;
	uid_t uid;
	gid_t gid;
	int wstat;
	int c;
	char *dir = NULL;
	struct stat sb;
	int vopt = 0;
	int popt = 0;
	int oopt = 0;
	
	prog = basename(argv[0]);
        while ((c = getopt_long(argc, argv, OPTIONS, longopts, NULL)) != EOF) {
                switch (c) {
			case 'o': /* --slash-opt */
				oopt++;
				break;
                        case 'd': /* --dir */
                                dir = optarg;
                                break;
                        case 'v': /* --verbose */
				vopt++;
				break;
                        case 'p': /* --preserve */
				popt++;
				break;
                        default:
                                usage();
                                break;
                }
        }
	if (optind == argc)
		usage();
	uid = getuid();
	gid = getgid();
	if (uid == 0) {
		fprintf(stderr, "%s: will not run as root\n", prog);
		exit(1);
	}

	if (dir || popt) {	/* no need to clean up so no fork */
		if (dir && (lstat(dir, &sb) < 0 || !S_ISDIR(sb.st_mode))) {
			fprintf(stderr, "%s: %s: not a directory\n", prog, dir);
			exit(1);
		}
		if (!dir && !(dir = create_tmpdir(uid, gid)))
			exit(1);
		exec_cmd(uid, gid, oopt, vopt, dir, &argv[optind]);
		exit_code = 1;
	} else {
		if (!(dir = create_tmpdir(uid, gid)))
			exit(1);
		switch ((pid = fork())) {
			case -1:	/* error */
				fprintf(stderr, "%s: fork: %m\n", prog);
				exit_code = 1;
			case 0:		/* child */
				exec_cmd(uid, gid, oopt, vopt, dir, &argv[optind]);
				exit(1);
			default:	/* parent */
				p = waitpid(pid, &wstat, 0);
				if ((int) p == -1) {
					fprintf(stderr, "%s: waitpid: %m\n", 
							prog);
					exit_code = 1;
				} else if (WIFEXITED(wstat)) {
					exit_code = WEXITSTATUS(wstat);
				} else { /* signal = error */
					exit_code = 1;
				}
				break;
		}
		if (!popt) {
			if (vopt)
				fprintf(stderr, "%s: removing %s\n", prog, dir);
			if (!recursive_remove(dir)) {
				fprintf(stderr, "%s: failed to clean up %s\n", 
					prog, dir);
				exit_code = 1;
			}
		}
	}
	exit(exit_code);
}

static void
exec_cmd(uid_t uid, gid_t gid, int oopt, int vopt, char *dir, char *argv[])
{
	char *target = oopt ? "/opt" : "/usr/local";

	if (unshare(CLONE_NEWNS) < 0) {
		fprintf(stderr, "%s: unshare: %m\n", prog);
		return;
	}
	if (vopt)
		fprintf(stderr, "%s: binding %s on %s\n", prog, dir, target);
	if (mount(dir, target, NULL, MS_BIND, NULL) < 0) {
		fprintf(stderr, "%s: mount: %m\n", prog);
		return;
	}
	if (vopt)
		fprintf(stderr, "%s: dropping root privileges\n", prog);
	if (setgid(gid) < 0) {
		fprintf(stderr, "%s: setgid: %m\n", prog);
		return;
	}
	if (setuid(uid) < 0) {
		fprintf(stderr, "%s: setuid: %m\n", prog);
		return;
	}
	if (vopt)
		fprintf(stderr, "%s: execing %s\n", prog, argv[0]);
	execvp(argv[0], argv);
	fprintf(stderr, "%s: exec: %s: %m\n", prog, argv[0]);
}

static int
recursive_remove(char *name)
{
	struct dirent *dp;
	DIR *dir; 
	int res = 1;
	char path[MAXPATHLEN];

	if (!(dir = opendir(name))) {
		fprintf(stderr, "%s: opendir %s: %m\n", prog, name);
		return 0;
	}
	while ((dp = readdir(dir))) {
		if (!strcmp(dp->d_name, "."))
			continue;
		if (!strcmp(dp->d_name, ".."))
			continue;
		snprintf(path, sizeof(path), "%s/%s", name, dp->d_name);
		if (dp->d_type == DT_DIR) { 
			if (!recursive_remove(path)) {
				res = 0;
				break;
			}
		} else {
			if (unlink(path) < 0) {
				fprintf(stderr, "%s: unlink %s: %m\n", prog, path);
				res = 0;
				break;
			}
		}
	}
	if (closedir(dir) < 0) {
		fprintf(stderr, "%s: closedir %s: %m\n", prog, name);
		res = 0;
	}
	if (rmdir(name) < 0) {
		fprintf(stderr, "%s: rmdir %s: %m\n", prog, name);
		res = 0;
	}
	return res;
}


static char *
create_tmpdir(uid_t uid, gid_t gid)
{
	static char template[] = "/tmp/tmplocal-XXXXXX";

	if (mkdtemp(template) == NULL) { /* mode 0700 */
		fprintf(stderr, "%s: mkdtemp failed\n", prog);
		return NULL;
	}
	if (chown(template, uid, gid) < 0) {
		fprintf(stderr, "%s: chown: %m\n", prog);
		recursive_remove(template);
		return NULL;
	}
	return template;
}

